File: Greedy/630-H-Course-Schedule-III/course_schedule_III.py
import heapq as hq
class Solution:
    def scheduleCourse(self, courses: list[list[int]]) -> int:
        courses.sort(key=lambda x:x[1]) # sort by lastDay (..<=.. <=..)

        total_time = 0
        my_lst: list[int] = []

        for course in courses:
            hq.heappush(my_lst, -course[0])
            total_time += course[0]
            if total_time > course[1]:
                total_time += hq.heappop(my_lst)
        return len(my_lst)

File: Greedy/630-H-Course-Schedule-III/course_schedule_iii.cpp
# include <vector>
# include <algorithm>
# include <queue>

class Solution {
public:
    int scheduleCourse(vector<vector<int>>& courses) {
        sort(courses.begin(), courses.end(), [](const vector<int>&a, const vector<int>&b) {return a<b;});
        priority_queue<int> maxheap;

        int total_time = 0;
        for (const auto& course : courses) {
            int duration = course[0];
            int lastDay = course[1];

            maxheap.push(duration);
            total_time += duration;

            if (total_time > lastDay) {
                total -= maxheap.top();
                maxheap.pop();
            }
            
        }
        return maxheap.size();
    }
};